lib/pg_party/adapter_decorator.rb

```diff
@@ -120,6 +255,81 @@ def attach_partition(parent_table_name, child_table_name, constraint_clause)
       PgParty.cache.clear!
     end
 
+    def recursive_add_index(table_name:, index_name:, index_type:, index_columns:, index_options:, using:, algorithm:,
+                            in_threads: nil, _parent_index_name: nil, _created_index_names: [])
+      partitions = partitions_for_table_name(table_name, include_subpartitions: false)
+      updated_name = _created_index_names.empty? ? index_name : generate_index_name(index_name, table_name)
+
+      # If this is a partitioned table, add index ONLY on this table.
+      if table_partitioned?(table_name)
+        add_index_only(table_name, type: index_type, name: updated_name, using: using, columns: index_columns,
+                       options: index_options)
+        _created_index_names << updated_name
+
+        parallel_map(partitions, in_threads: in_threads) do |partition_name|
+          recursive_add_index(
+            table_name: partition_name,
+            index_name: index_name,
+            index_type: index_type,
+            index_columns: index_columns,
+            index_options: index_options,
+            using: using,
+            algorithm: algorithm,
+            _parent_index_name: updated_name,
+            _created_index_names: _created_index_names
+          )
+        end
+      else
+        _created_index_names << updated_name # Track as created before execution of concurrent index command
+        add_index_from_options(table_name, name: updated_name, type: index_type, algorithm: algorithm, using: using,
+                               columns: index_columns, options: index_options)
+      end
+
+      attach_child_index(updated_name, _parent_index_name) if _parent_index_name
+
+      return true if index_valid?(updated_name)
```

Comment:
I was under the impression that AR would raise an error when index creation fails (even using the concurrent option). What's the reasoning for explicitly validating the index?

Reply:
The way that Postgres 11+ indices work on partitioned tables let me to this solution to ensure the whole operation succeeds. The process is as follows:
1. You create an index ON ONLY the parent (partitioned) table - this index is immediately marked invalid by postgres
2. You create all indices on all child partitions, repeating 1 recursively for subpartitions.
  a. After creating each index you ATTACH it to the parent index
3. Once all partitions have indices which are attached to the parent index, Postgres detects this and marks the parent index as valid. If you miss any partitions, the parent index is left in an invalid state and not used by the query planner

So basically, the validity check needed to ensure the entire hierarchy of indices was successful created and attached up the chain. If anything was missed, the validity check will fail and the method with unwind all the indices created via DROP INDEX IF EXISTS so it doesn't leave any orphaned indices.

Because the whole operation is done outside of a transaction when using CONCURRENTLY and also the support for parallel threads, I felt like this was an important check to add especially.